### ISSUE TITLE
[feature request] Extract class to a new file

### ISSUE THREAD

**Issue by JFixby** [dart-lang/sdk#30310]
![image](https://user-images.githubusercontent.com/1580663/28843411-5b3a1b16-7701-11e7-9db2-3c7e76c0c99c.png)

**Comment by scullabyte** [comment]
Any update on this? Possibly a way to do this with another application that will update all references? Thanks!

**Comment by cnsumner** [comment]
This would be extremely useful in VSCode. I have to do this operation manually 10s of times while mocking out an app UI.

**Comment by AndroidNils** [comment]
Any updates?

**Comment by bwilkerson** [comment]
Unfortunately, no. I think there's a lot of value to this and similar refactorings that aid in code re-organization, but this work hasn't yet been prioritized high enough to be scheduled.

**Comment by magnussp** [comment]
Would love to see this feature added. Makes life so much easier.

**Comment by ajaygautam** [comment]
This has 86 likes now... in case you are using that as a voting system. (PS: I have no idea what other issues are that have higher likes/votes)

@bwilkerson code re-organization is a programmer productivity task. I would very much love to see things that will make me more productive. Hence I disagree on your lack of value stance. Imagine the combined effort spent in manual refactoring!

**Comment by keehoo** [comment]
almost 4 years... geez..

**Comment by dkbast** [comment]
Looks like Jetbrains has this in Rider for .net ... I keep ending up on this site every few months when looking for this feature so I thought I'd share it here:
https://www.jetbrains.com/help/rider/Refactorings__Move__Type_to_Another_File.html

**Comment by dfaveur** [comment]
Yes please. It is very cumbersome currently to isolate widget per file

**Comment by krema** [comment]
I just discovered a plugin that we can use in the meantime.

https://github.com/marius-h/flutter_enhancement_suite

![Screen Shot 2021-03-13 at 14 36 32 PM](https://user-images.githubusercontent.com/1336504/111031802-9a1b8900-8409-11eb-8ae5-ce8dea13be97.png)

**Comment by luca-venir** [comment]
Unluckily such extension doesn't work on VS code.

**Comment by asashour** [comment]
https://dart-review.googlesource.com/c/sdk/+/212562

**Comment by mrRedSun** [comment]
Why isn't this done yet?

**Comment by scheglov** [comment]
Works for me.

1. Completion.
<img width="750" alt="image" src="https://user-images.githubusercontent.com/384794/180091130-aa348f26-aeb6-4e0a-bda9-e80bc20d330a.png">

2. Quick fix.
<img width="541" alt="image" src="https://user-images.githubusercontent.com/384794/180091183-fe493dde-8395-4ffe-bd99-a95593a04057.png">

**Comment by scheglov** [comment]
@srawlins I was talking only about extensions https://github.com/dart-lang/sdk/issues/38894, nothing about this one :-)

**Comment by srawlins** [comment]
Ah exellent. Thanks!

**Comment by mannyvergel** [comment]
Belated happy 5th birthday to this issue 🥳

**Comment by dhxyz** [comment]
Are there any updates on this? [#1831 ](https://github.com/Dart-Code/Dart-Code/issues/1831) seems promising.

**Comment by maks** [comment]
No its the other way around, the the DartCode issue 1831 can only go ahead once its implemented here in the analyzer. 

https://dart-review.googlesource.com/c/sdk/+/212562 is over a year old, @bwilkerson  did it get abandoned? It doesn't look like it ever got reviewed?

**Comment by dhxyz** [comment]
Alright, I understand now. Thank you for explaining. ❤️

**Comment by bwilkerson** [comment]
> https://dart-review.googlesource.com/c/sdk/+/212562 is over a year old, @bwilkerson did it get abandoned? It doesn't look like it ever got reviewed?

I'm not sure why there are no comments on it, but Ahmed and I did have a conversation about it. The conversation should have been on the issue for posterity.

The problem with that CL is that, while it implements the refactoring on the server side, none of the clients would have support for it. Our refactoring support is too manual at the moment, requiring work on every individual client in order to make a refactoring available (details omitted for brevity). That isn't a sustainable approach. I wanted to fix that problem before adding any new refactorings. If I had known how long it was going to take to make that happen I might have made a different choice at the time.

The good news is that we are (slowly, unfortunately) working on adding support for a new refactoring protocol that is much more flexible. It allows clients to add support for refactorings in general and then for server to send a list of available refactorings at the time the client requests them.

The new approach looks promising, and the "move a top level declaration to a file" refactoring is the first proof of concept refactoring that we're implementing. I don't have an estimate for when it will be ready to ship, but I do plan to continue to push it forward. Once the general mechanism is completed we should be able to add refactorings to the language server without needing to update the clients, making it much faster to implement new refactorings than it has been in the past.

**Comment by mattfennerjesi** [comment]
Thanks @bwilkerson that sounds very promising. Is there another issue we can watch for progress on this refactor?

**Comment by bwilkerson** [comment]
No, I think this one is sufficient.

At this point there are a couple of rough edges, primarily related to import prefixes and part files, but the core functionality is implemented (though the functionality is not yet enabled).

**Comment by dhxyz** [comment]
> No, I think this one is sufficient.
> 
> At this point there are a couple of rough edges, primarily related to import prefixes and part files, but the core functionality is implemented (though the functionality is not yet enabled).

That's great news, thank you so much! Soon, I can finally take care of the many thousand lines long spaghetti files in my project. 😅

Is there any way to access this feature in its current state?

**Comment by mika76** [comment]
Could someone point me to the hidden flag? I'd love to try and test it out...

**Comment by DanTup** [comment]
The flag goes in your VS Code settings and is:

```js
"dart.experimentalNewRefactors": true,
```

As Brian noted above, the code that shipped in the latest stable release is not complete and may have issues, so I also wouldn't recommend using this on that version (it should be more complete in the [Dart 3 alpha](https://medium.com/dartlang/dart-3-alpha-f1458fb9d232)).

Note that VS will fade the setting above out as if it's an invalid setting because it's not declared in the manifest. This setting may also change or be reused in future (to aid testing of other new refactors), as it is currently the flag to enable the new refactor functionality, rather than this specific refactor.

**Comment by mika76** [comment]
@DanTup much appreciated

**Comment by Musta-Pollo** [comment]
Great job! Thank you.

Currently, it just puts imports at the bottom of the page.

**Comment by DanTup** [comment]
@Musta-Pollo that should be fixed by https://github.com/dart-lang/sdk/commit/12b606e9e95cc6622ba2230a184c3a068284ec6a, but it was only a few weeks ago so perhaps not in the version you're testing.

**Comment by Musta-Pollo** [comment]
Thanks, on master, it works correctly.

**Comment by cbenhagen** [comment]
@DanTup do you know if this can already be tested in the Intellij plugin? Haven't found a related issue in the plugin repo.

**Comment by DanTup** [comment]
@cbenhagen as far as I know, the new refactor system is only currently available over LSP (the newer protocol used by VS Code).

For IntelliJ, it'll need adding to the original server protocol (and work in the plugin to handle the server asking for input like filenames), but I don't think work has started on that yet.

**Comment by DanTup** [comment]
FYI, I'm making some slight changes to the opt-in flag for experimental refactors. The existing flag (`"dart.experimentalNewRefactors"`) was for entirely enabling the new refactoring system. This was fine for initial development but doesn't allow us to have both experimental + stable refactors using this system in future.

In an upcoming release, that flag will be replaced by `"dart.experimentalRefactors"` (the word "New" has been removed) and it will now control only specific refactors that are marked as experimental.

For the time being, the "Move to File" refactor will continue to be considered experimental. You're welcome to try it out and provide feedback to help find any remaining issues, but be aware that in the current stable and current Dart 3 pre-release SDKs it's incomplete and has several known issues, such as:

- file headers may be inserted after the moved class
- moved classes may be inserted above import statements
- moving sealed classes does not automatically move their direct subclasses
- moving sealed classes/subclasses that are spread across `part` files is allowed and results in invalid code (for now, moving sealed types with direct subclasses in other `part`s will not be supported)

If you find any issues _not_ listed above, please file new issues (and feel free to cc me). If you want to update your settings in advance so the experimental refactor doesn't disappear when you get an SDK with this change, you can include both flags in your VS Code settings:

```js
	"dart.experimentalNewRefactors": true,
	"dart.experimentalRefactors": true,
```

Though be aware that these flags may also cause other incomplete refactors to show up in future, so you may wish to disable them once this refactor is no longer considered experimental.

**Comment by dhxyz** [comment]
@DanTup:

Thank you so much for implementing this feature. I have been using it quite extensively in the meantime and it has saved me a lot of time.

I would like to propose something for the imports, though. Instead of, for example, importing material.dart, the feature imports material widgets selectively, like so:

```
import 'package:flutter/src/material/icons.dart';
import 'package:flutter/src/material/page.dart';
import 'package:flutter/src/painting/border_radius.dart';
import 'package:flutter/src/painting/box_border.dart';
import 'package:flutter/src/painting/box_decoration.dart';
import 'package:flutter/src/painting/edge_insets.dart';
import 'package:flutter/src/rendering/box.dart';
import 'package:flutter/src/rendering/flex.dart';
import 'package:flutter/src/widgets/basic.dart';
import 'package:flutter/src/widgets/container.dart';
import 'package:flutter/src/widgets/framework.dart';
import 'package:flutter/src/widgets/icon.dart';
import 'package:flutter/src/widgets/navigator.dart';
import 'package:flutter/src/widgets/text.dart';
```

To prevent this, I have always deleted all of the automatic imports and then I just pasted all of the imports of the source file at the top of the new file. Then, I hit save, which triggers "fix all" through codeActionsOnSave. This removes all unused imports.

Please bear in mind that I am a total noob in asking this, but: Wouldn't it be easier and better if experimentalRefactors did the same thing as I do manually (pasting all imports from the source file & fix all) instead of looking for every widget selectively (if that even is what it does)?

I am still on Dart 2.19.6, though. So, this might not even be relevant anymore. 😅 Thanks, again. It's an awesome feature.

**Comment by DanTup** [comment]
@dhxyz FWIW it's discouraged to import from another packages `src` folder - see https://dart.dev/guides/language/effective-dart/usage#dont-import-libraries-that-are-inside-the-src-directory-of-another-package.

> Wouldn't it be easier and better if experimentalRefactors did the same thing as I do manually (pasting all imports from the source file & fix all) instead of looking for every widget selectively (if that even is what it does)?

If I understand correctly, the code essentially does an improved version of this. It locates all of the _relevant_ imports and then re-imports them into the destination file (with the same prefixes if applicable).

However, when imports are added there is some code that tries to "improve" imports if they are in a `src` folder that you shouldn't be importing from (this fixes issues like #49081). I guess there's a question here of whether that code should be running in this case.

@bwilkerson do you have an opinion on what we should do if the source file contains `src` imports? One one hand this seems desired (we don't want to add `src` imports where they shouldn't be used), but on the other hand if a user already had these imports they might expect it to copy them.

**Comment by dhxyz** [comment]
> do you have an opinion on what we should do if the source file contains src imports?

About that - this is not what is happening in my case. Not sure if I expressed that correctly, sorry for that. The above example is what the feature does when the source file (by that I mean the file where the widget I want to refactor is from) contains 'package:flutter/material.dart' as an import. There were no src imports in the source file.

**Comment by DanTup** [comment]
@dhxyz oh sorry, I mis-read what you wrote. If it's adding `package:flutter/src` imports that weren't in your source file that's definitely a bug (and sounds like it's _not_ using the code I blamed above). I'm out today, but will check if this still occurs on the latest code when I'm back next week. Thanks!

**Comment by bwilkerson** [comment]
I agree, that's definitely a bug. Thanks for bringing this to our attention so that we can get it fixed before shipping the refactoring.

**Comment by babbage** [comment]
This is great, most useful. I likewise am seeing the bug reported above where imports from the flutter framework are over-specific instead of just `import 'package:flutter/material.dart';`

I also noted another issue with our code I wanted to report. In the original file:
`import 'dart:developer' as developer;`

In the new file from the refactoring:
`import 'dart:developer';`

As a result, calls in the new file to `developer.log(...)` are not resolved correctly and are marked as errors.

**Comment by DanTup** [comment]
@babbage I can reproduce that (it specifically seems to be that prefixed functions are not handled) and have a fix in progress. Thanks!

**Comment by mihalycsaba** [comment]
Enabled it and it works well enough, it's slower than other refactor actions, takes like 2 seconds, but it's okay better than doing this manually.

One suggestion for improvement if it's possible. When the code was moved to a new file, the original file still retained the imports that were only used by that code. It would be nice if it removed the imports that are no longer needed.

**Comment by bwilkerson** [comment]
We'll have to look into the performance issue to see whether it's possible to improve the situation, but thanks for letting us know.

Yes, not removing unused imports from the original file is a known issue. I agree it would be nice for the refactoring to do that for you. We decided to ship without it because we have a separate fix to remove unused imports (a workaround) and thought it was better to get the refactoring out to users sooner.

**Comment by mihalycsaba** [comment]
I think this should work for functions too.

**Comment by bwilkerson** [comment]
As it's currently implemented, it ought to work for any set of contiguous non-private top-level declarations (though there might be a couple of exceptions I'm forgetting about). If it isn't working for some particular case, please open a new issue and provide a reproduction case.

**Comment by Autom8r-82** [comment]
@bwilkerson Is it possible to move to an existing file without it being overwritten? I'd like to consolidate Classes and Methods into an existing file but it appears that the file will always be overwritten on each refactor. Am I missing something?

**Comment by DanTup** [comment]
Moving to an existing file is supported.. however I just tested it and see that when I select an existing file VS Code gives this message:

![image](https://github.com/dart-lang/sdk/assets/1078012/88fffa70-854c-491a-a2b5-5df8711204fb)

This message is misleading, and we actually do not replace the file, we merge into it. It seems like it's because of the VS Code API we use to select a file (showSaveDialog). I'm not sure if it's always done and I missed it, or it's new.

I've opened https://github.com/microsoft/vscode/issues/203326 in VS Code to see if there could be a way to suppress this (or have an API for selecting files that does not do this).

**Comment by Autom8r-82** [comment]
@DanTup thanks for the insight! I stopped dead in my tracks after receiving the new, rather, scary looking warning in VS Code:

![image](https://github.com/dart-lang/sdk/assets/149018152/a006f15e-4dd6-41a3-991d-75672a562aa7)

I'll have to give this a go myself, it'll be very useful indeed!

**Comment by DanTup** [comment]
@Autom8r-82 oof, that one is even worse (it more explicitly states that something bad will happen that will not). Thanks for the heads up - I'll see if I can find any other options for this.

**Comment by FMorschel** [comment]
Just to point out the importing of necessary libraries could be handled by https://github.com/dart-lang/sdk/issues/30296.

**Comment by bwilkerson** [comment]
I was under the impression that imports are correctly being computed. Is that not the case?

**Comment by FMorschel** [comment]
I missed that there was an existing flag to enable it. Sorry 🙃!

**Comment by DanTup** [comment]
> I missed that there was an existing flag to enable it

A flag to enable what? My understanding is that (for VS Code) this refactor is available on the latest SDKs and will add any necessary imports. No flags should be required (since [v3.72.0 of Dart-Code](https://dartcode.org/releases/v3-72/) was released and the stable SDK available at the time). If that's not what you're seeing, please let me know / file an issue.

(there are some issues with the dialog showing an overwrite prompt noted above which are not resolved yet though)

**Comment by DanTup** [comment]
@FMorschel sorry for the confusion, you're right that I didn't post in this issue when this refactor was marked as not-experimental (which I did in https://github.com/dart-lang/sdk/commit/a3399e3616f7040725bf943e8abe145c4a8e1108), only in the Dart-Code issue/release notes. I certainly should've noted it here.

Currently this refactor is only available in VS Code and not in IntelliJ/Android Studio. Making it available to IntelliJ/AS will require work in their Dart plugin, but might also require some additional work here to allow them to call the new refactoring system (possibly through LSP).

**Comment by bwilkerson** [comment]
@jwren

**Comment by theashggl** [comment]
I just landed on this issue and I think there should also be the feature available to not just extract class to ao new file but also extract any member of the class to another class and update the usages of that member accordingly with the object of the target class.

**Comment by bwilkerson** [comment]
> ... extract any member of the class to another class and update the usages of that member accordingly with the object of the target class.

This deserves an issue of its own. Would you mind opening one? And perhaps expand, in the new issue, on what you're looking for. For example, are you thinking of something that would move members up or down in the class hierarchy, or are you thinking more generally about moving to an arbitrary class? Would it be limited to existing classes, or are you wanting it to create a new class? What are some examples of what you mean by "update the usages"? What should it do about references to other members that are no longer visible? How should references to the moved member be updated?

**Comment by FMorschel** [comment]
> are you thinking of something that would move members up or down in the class hierarchy

I'm not exactly sure, but I think I saw an issue for this once, so maybe look for that too.
